keras_nlp/applications/bert.py

```diff
@@ -0,0 +1,372 @@
+        inner_activation="gelu",
+        initializer="glorot_uniform",
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self.embedding_table = embedding_table
+        self.inner_activation = keras.activations.get(inner_activation)
+        self.initializer = initializer
+
+    def build(self, input_shape):
+        self._vocab_size, hidden_size = self.embedding_table.shape
+        self.dense = keras.layers.Dense(
+            hidden_size,
+            activation=self.inner_activation,
+            kernel_initializer=self.initializer,
+            name="transform/dense",
+        )
+        self.layer_norm = keras.layers.LayerNormalization(
+            axis=-1, epsilon=1e-12, name="transform/LayerNorm"
+        )
+        self.bias = self.add_weight(
+            "output_bias/bias",
+            shape=(self._vocab_size,),
+            initializer="zeros",
+            trainable=True,
+        )
+
+        super().build(input_shape)
+
+    def call(self, sequence_data, masked_positions):
+        masked_lm_input = self._gather_indexes(sequence_data, masked_positions)
+        lm_data = self.dense(masked_lm_input)
+        lm_data = self.layer_norm(lm_data)
+        lm_data = tf.matmul(lm_data, self.embedding_table, transpose_b=True)
+        logits = tf.nn.bias_add(lm_data, self.bias)
+        masked_positions_length = (
+            masked_positions.shape.as_list()[1] or tf.shape(masked_positions)[1]
+        )
+        return tf.reshape(
+            logits, [-1, masked_positions_length, self._vocab_size]
+        )
+
+    def _gather_indexes(self, sequence_tensor, positions):
+        """Gathers the vectors at the specific positions, for performance.
+
+        Args:
+            sequence_tensor: Sequence output of shape
+                (`batch_size`, `seq_length`, `hidden_size`) where `hidden_size`
+                is number of hidden units.
+            positions: Positions ids of tokens in sequence to mask for
+                pretraining of with dimension (batch_size, num_predictions)
+                where `num_predictions` is maximum number of tokens to mask out
+                and predict per each sequence.
+
+        Returns:
+            Masked out sequence tensor of shape (batch_size * num_predictions,
+            `hidden_size`).
+        """
+        sequence_shape = tf.shape(sequence_tensor)
+        batch_size, seq_length = sequence_shape[0], sequence_shape[1]
+        width = sequence_tensor.shape.as_list()[2] or sequence_shape[2]
+
+        flat_offsets = tf.reshape(
+            tf.range(0, batch_size, dtype=tf.int32) * seq_length, [-1, 1]
+        )
+        flat_positions = tf.reshape(positions + flat_offsets, [-1])
+        flat_sequence_tensor = tf.reshape(
+            sequence_tensor, [batch_size * seq_length, width]
+        )
+        output_tensor = tf.gather(flat_sequence_tensor, flat_positions)
+
+        return output_tensor
+
+
+class BertLanguageModel(keras.Model):
```

Comment:
I don't think we should ship this right now for a couple reasons:

First (and most important), we aren't in a place where we can ship the preprocessing to go with pretraining right now, that will still need to live in the example code. So we are only solving half of a user journey here, preprocessing will still need to be copied from a template. We could try to ship the preprocesssing for pretraining, but there's actually some significant design Qs there (can we do BERT pretraining with a tf.data pipeline??).

Second (and much more easily fixed), this is shipping what is basically a half compiled model. We set up a loss and metrics, yet no optimizer and learning rate. This is an unexpected thing to deliver to users. We need to plan out the whole expected usage of this thing, or just deliver the architecture all the way to logits, but let users fully compile the model themselves.

If we are shipping this, I would suggest not doing in the first version, and actually starting with RoBERTa here, where we can easily the preprocessing for pretraining along with the model and deliver a whole user journey.

Reply:
Then there's an even more zoomed out question of do we want to ship an API including compilation settings for pretraining, or should we just demo copy-pastable code there? I will pull that question into the doc.

Reply:
I agree and this is a great way to scope down the PR, which is a bit too ambitious for my taste.